src/main/java/com/rabbitmq/jms/client/RMQSession.java

```diff
@@ -140,6 +146,9 @@
      * we must never acknowledge a message more than once (nor acknowledge a message that doesn't exist). */
     private final SortedSet<Long> unackedMessageTags = Collections.synchronizedSortedSet(new TreeSet<Long>());
 
+    /* Holds the uncommited tags to commit a nack on rollback */
+    private final ArrayList<Long> uncommittedMessageTags = new ArrayList<Long>(); 
```

Comment:
The property should always be declared with the interface (e.g. `List`), not with the implementation. This must be synchronized. `Collections.synchronizedList` should do the trick.

Reply:
Right. It'd be good to add `// GuardedBy("commitLock");` just after the declaration, to make it clear that this data structure must be synchronized explicitly.